### PR TITLE
feat: Add topological sort and disjoint set algorithms

### PR DESCRIPTION
**Pull request by Siddhram**
## Topological Sort (Kahn's Algorithm)
Finds a linear ordering of vertices in a directed acyclic graph (DAG).  
Detects cycles and resolves dependencies efficiently.  
Time Complexity: O(V + E)  
Space Complexity: O(V) for adjacency list and queue.  
Includes examples for task scheduling and dependency resolution.

## Disjoint Set (Union-Find)
Efficiently performs set operations using path compression and union by rank.  
Supports find_set and union_sets operations for connected components and network connectivity.  
Time Complexity: Near O(1) amortized per operation  
Space Complexity: O(N) for parent and rank arrays.  
Includes examples demonstrating graph connectivity and merging operations.

Both implementations include:  
- Clear documentation and real-world applications  
- Multiple test cases and edge case handling  
- Optimized performance and readable code structure

### CODE REVIEW

File: graph_algorithms/disjoint_set.r [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  #' Uses path compression for efficiency
+  #' @param ds: Disjoint set data structure
+  #' @param v: Element to find
+  #' @return: Representative of v's set
+  
+  if (ds$parent[v + 1] != v) {
+    # Path compression: Make all nodes on path point to root
+    ds$parent[v + 1] <- find_set(ds, ds$parent[v + 1])
+  }
+  ds$parent[v + 1]
```
[the remark this answers is not in the document]
Comment (siriak) [reply graph_algorithms/disjoint_set.r:41]: Please check

File: graph_algorithms/disjoint_set.r [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+# Test case 2: Path compression
+cat("\n=== Test Case 2: Path Compression ===\n")
+ds2 <- make_set(6)
+ds2 <- union_sets(ds2, 0, 1)
+ds2 <- union_sets(ds2, 1, 2)
+ds2 <- union_sets(ds2, 2, 3)
+cat("Parents before find:", paste(ds2$parent, collapse = " "), "\n")
+find_set(ds2, 3)  # This should compress the path
+cat("Parents after find:", paste(ds2$parent, collapse = " "), "\n")
```
[the remark this answers is not in the document]
Comment (siriak) [reply graph_algorithms/disjoint_set.r:95]: Please check
